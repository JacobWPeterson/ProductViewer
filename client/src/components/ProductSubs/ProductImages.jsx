import React from 'react';
import styled from 'styled-components';
import FontAwesome from 'react-fontawesome';

import ViewerThumbnails from './ViewerThumbnails.jsx';
import ExpandedImage from './ExpandedImage.jsx';

const Wrapper = styled.div`
  position: relative;
  flex-basis: 65%;
  width: 50vw;
  height: 70vh;
  display: flex;
  flex-direction: row;
  border: 2px solid #aeaeae;
`;

const RightArrow = styled.div`
  color: #e0e0e0;
  font-size: 1.75rem;
  z-index: 1;
  position: absolute;
  top: 30vh;
  right: .25vw;
  &:hover {
    color: #80ccc4;
    transform: scale(1.1);
  };
`;

const LeftArrow = styled.div`
  color: #e0e0e0;
  font-size: 1.75rem;
  z-index: 1;
  position: absolute;
  top: 30vh;
  left: git.25vw;
  &:hover {
    color: #80ccc4;
    transform: scale(1.1);
  };
`;

const Image = styled.img`
  width: 100%;
  height: 100%;
  object-fit: cover;
  padding: 0;
  position: relative;
  z-index: 0;
  &:hover { cursor: crosshair; };
`;

const ModalBackground = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 4;
  background-color: #42424275;
`;

class ProductImages extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      index: 0,
      start: 0,
      end: 4,
      isModalOpen: false,
    };
    this.clickNavHandler = this.clickNavHandler.bind(this);
    this.clickedThumb = this.clickedThumb.bind(this);
    this.indexUpdater = this.indexUpdater.bind(this);
    this.indexChecker = this.indexChecker.bind(this);
    this.toggleModal = this.toggleModal.bind(this);
  }

  clickNavHandler(event) {
    const direction = Number(event.target.id);
    if (this.state.index + direction >= 0
      && this.state.index + direction < this.props.images.length) {
      this.setState((prevState) => ({ index: prevState.index + direction }),
        () => this.indexChecker());
    }
  }

  clickedThumb(index) {
    this.setState({ index: Number(index) });
  }

  indexUpdater(amount) {
    if (this.state.start + amount >= 0
      && this.state.end + amount < this.props.images.length) {
      this.setState((prevState) => ({
        start: prevState.start + amount,
        end: prevState.end + amount,
      }));
    }
  }

  indexChecker() {
    if (this.state.index < 5) {
      this.setState({
        start: 0,
        end: 4,
      });
    } else if (this.props.images.length - this.state.index <= 5) {
      this.setState({
        start: this.props.images.length - 5,
        end: this.props.images.length - 1,
      });
    } else if (this.state.index < this.state.start) {
      this.setState((prevState) => ({
        start: prevState.index,
        end: prevState.index + 4,
      }));
    } else if (this.state.index > this.state.end) {
      this.setState((prevState) => ({
        start: prevState.index - 4,
        end: prevState.index,
      }));
    }
  }

  toggleModal() {
    this.setState((prevState) => ({
      isModalOpen: !prevState.isModalOpen,
    }));
  }

  render() {
    // console.log(this.props.images)
    return (
      <Wrapper>

        {this.state.index !== 0 && <LeftArrow onClick={this.clickNavHandler}><FontAwesome id="-1" name="angle-left" size="2x" /></LeftArrow> }

        <ViewerThumbnails
          viewerIndex={this.state.index}
          start={this.state.start}
          end={this.state.end}
          indexUpdater={this.indexUpdater}
          images={this.props.images}
          clickedThumb={this.clickedThumb}
          id={this.props.id}
          alt=""
        />

        <Image onClick={this.toggleModal} src={this.props.images[this.state.index].url} key={this.props.id} alt="style photograph" />

        {this.state.isModalOpen && (
          <ModalBackground onMouseDown={this.toggleModal}>
            <ExpandedImage
              src={this.props.images[this.state.index].url}
              id={this.props.id}
              alt="style photograph"
              isModalOpen={this.state.isModalOpen}
              toggleModal={this.toggleModal}
              index={this.state.index}
              images={this.props.images}
              clickNavHandler={this.clickNavHandler}
              clickedThumb={this.clickedThumb}
            />
          </ModalBackground>
        )}

        {this.state.index !== this.props.images.length - 1 && <RightArrow onClick={this.clickNavHandler}><FontAwesome id="1" name="angle-right" size="2x" /></RightArrow> }

      </Wrapper>
    );
  }
}

export default ProductImages;
