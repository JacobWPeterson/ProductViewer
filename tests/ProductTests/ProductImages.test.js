/* eslint-disable no-undef */
/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount } from 'enzyme';

import ProductImages from '../../client/src/components/ProductSubs/ProductImages.jsx';
import ProductTestData from '../ProductTestData.js';

const props = {
  index: 0,
  images: ProductTestData.productStyles.results[0].photos,
};

describe('ProductImages', () => {
  const ProductImagesComponent = mount(<ProductImages {...props} debug />);

  it('Should render correctly', () => {
    expect(ProductImagesComponent).toMatchSnapshot();
  });
});
