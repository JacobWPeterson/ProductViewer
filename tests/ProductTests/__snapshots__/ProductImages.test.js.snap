// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImages Should render correctly 1`] = `
<ProductImages
  debug={true}
  images={
    Array [
      Object {
        "thumbnail_url": "https://images.unsplash.com/photo-1552902865-b72c031ac5ea?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1552902865-b72c031ac5ea?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
      },
      Object {
        "thumbnail_url": "https://images.unsplash.com/photo-1492447105260-2e947425b5cc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1492447105260-2e947425b5cc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80",
      },
      Object {
        "thumbnail_url": "https://images.unsplash.com/photo-1548133464-29abc661eb5c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1548133464-29abc661eb5c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
      },
      Object {
        "thumbnail_url": "https://images.unsplash.com/photo-1500340520802-1687634cbe38?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1500340520802-1687634cbe38?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
      },
      Object {
        "thumbnail_url": "https://images.unsplash.com/photo-1559304022-afbf28f53c4d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1559304022-afbf28f53c4d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1656&q=80",
      },
      Object {
        "thumbnail_url": "https://images.unsplash.com/photo-1554921148-83d8ceda2095?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1554921148-83d8ceda2095?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
      },
    ]
  }
  index={0}
>
  <ProductImages__Wrapper>
    <div
      className="ProductImages__Wrapper-sc-1k36oai-0 ffMCcO"
    >
      <ViewerThumbnails
        alt=""
        clickedThumb={[Function]}
        end={4}
        images={
          Array [
            Object {
              "thumbnail_url": "https://images.unsplash.com/photo-1552902865-b72c031ac5ea?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1552902865-b72c031ac5ea?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
            },
            Object {
              "thumbnail_url": "https://images.unsplash.com/photo-1492447105260-2e947425b5cc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1492447105260-2e947425b5cc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80",
            },
            Object {
              "thumbnail_url": "https://images.unsplash.com/photo-1548133464-29abc661eb5c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1548133464-29abc661eb5c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
            },
            Object {
              "thumbnail_url": "https://images.unsplash.com/photo-1500340520802-1687634cbe38?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1500340520802-1687634cbe38?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
            },
            Object {
              "thumbnail_url": "https://images.unsplash.com/photo-1559304022-afbf28f53c4d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1559304022-afbf28f53c4d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1656&q=80",
            },
            Object {
              "thumbnail_url": "https://images.unsplash.com/photo-1554921148-83d8ceda2095?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
              "url": "https://images.unsplash.com/photo-1554921148-83d8ceda2095?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
            },
          ]
        }
        indexUpdater={[Function]}
        start={0}
        viewerIndex={0}
      >
        <ViewerThumbnails__Thumbs>
          <div
            className="ViewerThumbnails__Thumbs-v6hnj6-0 kgjjgM"
          >
            <ViewerThumbnails__NoArrow>
              <div
                className="ViewerThumbnails__NoArrow-v6hnj6-6 hGnLFc"
              >
                <FontAwesome
                  id="-1"
                  name="angle-up"
                  size="2x"
                >
                  <span
                    aria-hidden={true}
                    className="fa fa-angle-up fa-2x"
                    id="-1"
                  />
                </FontAwesome>
              </div>
            </ViewerThumbnails__NoArrow>
            <div
              key="https://images.unsplash.com/photo-1552902865-b72c031ac5ea?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80"
            >
              <ViewerThumbnails__ViewerImageContainer>
                <div
                  className="ViewerThumbnails__ViewerImageContainer-v6hnj6-2 kCVtBt"
                >
                  <ViewerThumbnails__Image
                    id={0}
                    onClick={[Function]}
                    src="https://images.unsplash.com/photo-1552902865-b72c031ac5ea?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80"
                  >
                    <img
                      className="ViewerThumbnails__Image-v6hnj6-4 hXQyqE"
                      id={0}
                      onClick={[Function]}
                      src="https://images.unsplash.com/photo-1552902865-b72c031ac5ea?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80"
                    />
                  </ViewerThumbnails__Image>
                </div>
              </ViewerThumbnails__ViewerImageContainer>
              <ViewerThumbnails__Highlight>
                <div
                  className="ViewerThumbnails__Highlight-v6hnj6-3 kmKkYK"
                />
              </ViewerThumbnails__Highlight>
            </div>
            <ViewerThumbnails__ImageContainer
              key="https://images.unsplash.com/photo-1492447105260-2e947425b5cc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80"
            >
              <div
                className="ViewerThumbnails__ImageContainer-v6hnj6-1 dkKLIo"
              >
                <ViewerThumbnails__Image
                  id={1}
                  onClick={[Function]}
                  src="https://images.unsplash.com/photo-1492447105260-2e947425b5cc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80"
                >
                  <img
                    className="ViewerThumbnails__Image-v6hnj6-4 hXQyqE"
                    id={1}
                    onClick={[Function]}
                    src="https://images.unsplash.com/photo-1492447105260-2e947425b5cc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80"
                  />
                </ViewerThumbnails__Image>
              </div>
            </ViewerThumbnails__ImageContainer>
            <ViewerThumbnails__ImageContainer
              key="https://images.unsplash.com/photo-1548133464-29abc661eb5c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80"
            >
              <div
                className="ViewerThumbnails__ImageContainer-v6hnj6-1 dkKLIo"
              >
                <ViewerThumbnails__Image
                  id={2}
                  onClick={[Function]}
                  src="https://images.unsplash.com/photo-1548133464-29abc661eb5c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80"
                >
                  <img
                    className="ViewerThumbnails__Image-v6hnj6-4 hXQyqE"
                    id={2}
                    onClick={[Function]}
                    src="https://images.unsplash.com/photo-1548133464-29abc661eb5c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80"
                  />
                </ViewerThumbnails__Image>
              </div>
            </ViewerThumbnails__ImageContainer>
            <ViewerThumbnails__ImageContainer
              key="https://images.unsplash.com/photo-1500340520802-1687634cbe38?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80"
            >
              <div
                className="ViewerThumbnails__ImageContainer-v6hnj6-1 dkKLIo"
              >
                <ViewerThumbnails__Image
                  id={3}
                  onClick={[Function]}
                  src="https://images.unsplash.com/photo-1500340520802-1687634cbe38?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80"
                >
                  <img
                    className="ViewerThumbnails__Image-v6hnj6-4 hXQyqE"
                    id={3}
                    onClick={[Function]}
                    src="https://images.unsplash.com/photo-1500340520802-1687634cbe38?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80"
                  />
                </ViewerThumbnails__Image>
              </div>
            </ViewerThumbnails__ImageContainer>
            <ViewerThumbnails__ImageContainer
              key="https://images.unsplash.com/photo-1559304022-afbf28f53c4d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1656&q=80"
            >
              <div
                className="ViewerThumbnails__ImageContainer-v6hnj6-1 dkKLIo"
              >
                <ViewerThumbnails__Image
                  id={4}
                  onClick={[Function]}
                  src="https://images.unsplash.com/photo-1559304022-afbf28f53c4d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80"
                >
                  <img
                    className="ViewerThumbnails__Image-v6hnj6-4 hXQyqE"
                    id={4}
                    onClick={[Function]}
                    src="https://images.unsplash.com/photo-1559304022-afbf28f53c4d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80"
                  />
                </ViewerThumbnails__Image>
              </div>
            </ViewerThumbnails__ImageContainer>
            <ViewerThumbnails__DownArrow
              onClick={[Function]}
            >
              <div
                className="ViewerThumbnails__DownArrow-v6hnj6-7 iBaJVw"
                onClick={[Function]}
              >
                <FontAwesome
                  id="1"
                  name="angle-down"
                  size="2x"
                >
                  <span
                    aria-hidden={true}
                    className="fa fa-angle-down fa-2x"
                    id="1"
                  />
                </FontAwesome>
              </div>
            </ViewerThumbnails__DownArrow>
          </div>
        </ViewerThumbnails__Thumbs>
      </ViewerThumbnails>
      <ProductImages__Image
        alt="style photograph"
        onClick={[Function]}
        src="https://images.unsplash.com/photo-1552902865-b72c031ac5ea?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80"
      >
        <img
          alt="style photograph"
          className="ProductImages__Image-sc-1k36oai-3 CDgDl"
          onClick={[Function]}
          src="https://images.unsplash.com/photo-1552902865-b72c031ac5ea?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80"
        />
      </ProductImages__Image>
      <ProductImages__RightArrow
        onClick={[Function]}
      >
        <div
          className="ProductImages__RightArrow-sc-1k36oai-1 hSttQB"
          onClick={[Function]}
        >
          <FontAwesome
            id="1"
            name="angle-right"
            size="2x"
          >
            <span
              aria-hidden={true}
              className="fa fa-angle-right fa-2x"
              id="1"
            />
          </FontAwesome>
        </div>
      </ProductImages__RightArrow>
    </div>
  </ProductImages__Wrapper>
</ProductImages>
`;
